plans/example-browser-node/browser/index.js

```diff
@@ -0,0 +1,105 @@
+const { chromium, firefox, webkit } = require('playwright')
+const { exit } = require('process')
+const { spawn } = require('child_process')
+
+const spawnServer = require('../server')
+
+;(async () => {
+  spawnServer(8080)
+
+  let browser
+  try {
+    const browserDebugPort = process.env.BROWSER_DEBUG_PORT || 9222
+
+    switch (process.env.BROWSER_KIND || 'chromium') {
+      case 'chromium':
+        console.log(`launching chromium browser with exposed debug port: ${browserDebugPort}`)
+        browser = await chromium.launch({
+          args: [
+            '--remote-debugging-address=0.0.0.0',
+            `--remote-debugging-port=${browserDebugPort}`
+          ]
+        })
+        break
+
+      case 'firefox':
+        const localBrowserDebugPort = Number(browserDebugPort) + 1
+        console.log(`launching firefox browser with exposed debug port: ${browserDebugPort} (local ${localBrowserDebugPort})`)
+        browser = await firefox.launch({
+          args: [
+            `-start-debugger-server=${localBrowserDebugPort}`
+          ]
+        })
+
+        console.log('launching tcp proxy to expose firefox debugger for remote access')
+        const tcpProxy = spawn(
+          'socat', [
+            `tcp-listen:${browserDebugPort},bind=0.0.0.0,fork`,
+            `tcp:localhost:${localBrowserDebugPort}`
+          ]
+        )
+        tcpProxy.stdout.on('data', (data) => {
+          console.log(`firefox debugger: tcpProxy: stdout: ${data}`)
+        })
+
+        tcpProxy.stderr.on('data', (data) => {
+          console.error(`firefox debugger: tcpProxy: stderr: ${data}`)
+        })
+
+        break
+
+      case 'webkit':
+        console.log('launching webkit browser (remote debugging not yet supported)')
+        browser = await webkit.launch()
+        break
+    }
+
+    const page = await browser.newPage()
+
+    page.on('console', (message) => {
+      const loc = message.location()
+      console.log(`[${message.type()}] ${loc.url}@L${loc.lineNumber}:C${loc.columnNumber}: ${message.text()} — ${message.args()}`)
+    })
+
+    console.log('prepare page window (global) environment')
+    await page.addInitScript((nodeEnv) => {
+      const env = {}
+      for (const k in nodeEnv) {
+        if (k.startsWith('TEST_')) {
+          env[k] = nodeEnv[k]
+        }
+      }
+      window.testground = { env }
+    }, process.env)
+
+    console.log('opening up testplan webpage on localhost')
+    await page.goto('http://127.0.0.1:8080')
+
+    // TODO: wait for the test to finish somehow :|
+
+    console.log('start browser exit process...')
+
+    if (process.env.HALT_BROWSER_ON_FINISH === 'true') {
```

Comment:
Why is this needed?

Reply:
It is not needed. However, it can be useful to be able to prevent the test from exiting upon failure / success, and instead allow it to be open. This to be able to connect to it for a remote debugging session and restart the test by refreshing the browser with your breakpoints now set.

Reply:
Got it, we have similar behavior with the `keep_containers` flag. Could you use this approach? Something like `KeepOpenedBrowser`?

Reply:
Sure, will look into it.

Reply:
For now I have already named it `TEST_KEEP_OPENED_BROWSERS`, but seems like a proper runner solution like `keepOpenedBrowser` variable seems better suited for once we have a `docker:browser`, and then we can also see how to integrate browser-only things into the local docker runner. Resolving it as such, but please re-open if you do not agree.